File: src/java/BubbleSort.java
/**
 * Copyright (C) 2016 Canux CHENG.
 * All right reserved.
 * Name: BubbleSort.java
 */

//package java;

/**
 * Bubble Sort.
 * @author Canux CHENG canuxcheng@gmail.com
 * @version V1.0.0.0
 * @since Fri 16 Sep 2016 01:48:32 AM EDT
 */
public class BubbleSort
{
    public static void main(String[] args)
    {
        int[] array = {74, 27, 96, -1, 0, 35};

        printArray(array);
        bubbleSort(array);
        printArray(array);

        printArray(array);
        bubbleSortDesc(array);
        printArray(array);
    }

    public static void printArray(int[] array)
    {
        for (int i = 0; i < array.length; i++)
        {
            if (i == array.length - 1)
                System.out.println(array[i]);
            else
                System.out.print(array[i] + ",");
        }
    }

    /**
     * From min to max.
     * @param array
     * @return array
     */
    public static int[] bubbleSort(int[] array)
    {
        for (int i = 0; i < array.length - 1; i++)
        {
            for (int j = 0; j < array.length - 1 - i; j++)
            {
                if (array[j] > array[j + 1])
                {
                    int temp = array[j];
                    array[j] = array[j + 1];
                    array[j + 1] = temp;
                }
            }
        }
        return array;
    }

    /**
     * From max to min.
     * @param array
     * @return array
     */
    public static int[] bubbleSortDesc(int[] array)
    {
        for (int i = 0; i < array.length - 1; i++)
        {
            for (int j = 0; j < array.length - 1 - i; j++)
            {
                if (array[j] < array[j+1])
                {
                    int temp = array[j];
                    array[j] = array[j+1];
                    array[j+1] = temp;
                }
            }
        }
        return array;
    }
}

File: algorithm/bubble_sort.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
bubble_sort.py for sort algorithm for bubble sort.

Copyright (C) 2017 Canux CHENG.
All rights reserved.

LICENSE GNU General Public License v3.0.

:author: Canux CHENG canuxcheng@gmail.com
:version: V1.0.0.0
:since: Wed 18 Jan 2017 02:52:27 AM EST

DESCRIPTION:
"""


def bubble_sort(seq):
    """Main function for bubble sort from min to max.

    :param seq: include string/tupple/list.
    :type seq: sequence.
    :return seq: return the result after bubble sort.
    :rtype: sequence.
    """
    length = len(seq)
    for i in range(length):
        for j in range(1, length-i):
            if seq[j] < seq[j-1]:
                seq[j-1], seq[j] = seq[j], seq[j-1]
    return seq


def bubble_sort_desc(seq):
    """Main function for bubble sort from max to min.

    :param seq: include string/tupple/list.
    :type seq: sequence.
    :return seq: return the result after bubble sort with descending.
    :rtype seq: sequence.
    """
    length = len(seq)
    for i in range(length):
        for j in range(1, length-i):
            if seq[j] > seq[j-1]:
                seq[j-1], seq[j] = seq[j], seq[j-1]
    return seq


if __name__ == "__main__":
    seq = [200, 59, -6, 67, 0, 18, 59]
    print("source seq: {}".format(seq))
    print("after bubble sort: {}".format(bubble_sort(seq)))
    print("after bubble sort desc: {}".format(bubble_sort_desc(seq)))
